### PR TITLE
Add parameter to revoke old refresh token upon issuing new

### PR DESCRIPTION
**Pull request by ctriant**
Introduce parameter `revoke_refresh_on_issue` in order to revoke the Refresh Token used to issue a new one.

### CODE REVIEW

File: src/oidcop/oidc/token.py [hunk 1]
```diff
@@ -280,6 +277,8 @@ def process_request(self, req: Union[Message, dict], **kwargs):
             _resp["id_token"] = _idtoken.value
 
         token.register_usage()
+        if kwargs.get("revoke_refresh_on_issue", False):
```
Comment (nsklikas) [inline src/oidcop/oidc/token.py:280]: I think that this should be on the class (for global configuration) and on the client (for per-client configuration). We shouldn't have to pass it to the function every time. Maybe keep this and add the other 2 methods?

Also I think `True` is a better default value.

File: src/oidcop/oidc/token.py [hunk 2]
```diff
@@ -242,12 +239,12 @@ def process_request(self, req: Union[Message, dict], **kwargs):
...
-        if "issue_refresh" in kwargs:
-            issue_refresh = kwargs["issue_refresh"]
-        else:
-            if "offline_access" in scope:
-                issue_refresh = True
+
+        issue_refresh = kwargs.get("refresh_token", False)
+        # The existence of offline_access scope overwrites issue_refresh
+        if "offline_access" in scope:
+            issue_refresh = True
```
Comment (nsklikas) [inline src/oidcop/oidc/token.py:246]: The logic is reversed here. `issue_refresh` used to override `offline_access`, which IMHO is the correct behavior. Maybe this is better:
```python
        issue_refresh = kwargs.get("refresh_token", None)
        # The existence of offline_access scope overwrites issue_refresh
        if issue_refresh is None and "offline_access" in scope:
            issue_refresh = True
```

**Review by rohe — APPROVED** [review]

File: src/oidcop/oidc/token.py [hunk 3]
```diff
@@ -81,12 +81,9 @@ def process_request(self, req: Union[Message, dict], **kwargs):
...
 
-        issue_refresh = False
-        if "issue_refresh" in kwargs:
-            issue_refresh = kwargs["issue_refresh"]
-        else:
-            if "offline_access" in grant.scope:
-                issue_refresh = True
+        issue_refresh = kwargs.get("issue_refresh", False)
+        if "offline_access" in grant.scope:
+            issue_refresh = True
```
Comment (nsklikas) [inline src/oidcop/oidc/token.py:86]: The logic is reversed here as well. `issue_refresh` used to override `offline_access`, which IMHO is the correct behavior. Maybe this is better:
```python
        issue_refresh = kwargs.get("refresh_token", None)
        # The existence of offline_access scope overwrites issue_refresh
        if issue_refresh is None and "offline_access" in scope:
            issue_refresh = True
```

File: src/oidcop/oidc/token.py [hunk 4]
```diff
@@ -281,6 +278,19 @@ def process_request(self, req: Union[Message, dict], **kwargs):
...
         token.register_usage()
 
+        if ("client_id" in req
+            and req["client_id"] in _context.cdb
+            and "revoke_refresh_on_issue" in _context.cdb[req["client_id"]]
+        ):
+            revoke_refresh = _context.cdb[req["client_id"]].get("revoke_refresh_on_issue")
+        else:
+            revoke_refresh =(
+                _mngr.conf["token_handler_args"]["refresh"]["kwargs"].get("revoke_refresh_on_issue", False)
```
Comment (nsklikas) [inline src/oidcop/oidc/token.py:288]: Why add it to the token handler? Do you think it's a better there than the token class? You could add it to the token endpoint (oidcop/{oauth2,oidc}/token.py:Token) and then we could have here:
```python
revoke_refresh = self.endpoint.revoke_refresh_on_issue
```
what do think about this?

Comment (ctriant) [reply src/oidcop/oidc/token.py:288]: I thought token handler as a more suitable place as it describes the behavior upon tokens, in contrast with token class that describe the actual token.

The truth is that in oauth2/token.py: https://github.com/IdentityPython/oidc-op/blob/66eb57b305258308365b5439bb1f41c32950018b/src/oidcop/oauth2/token.py#L346 there is an unused parameter `new_refresh_token` that it seems to resembles the same functionality as `revoke_refresh`.

So I have no reason to insist on this decision.

**Review by peppelinux — APPROVED** [review]
I like this feature, thank you guys!
@ctriant let us know when this PR would be ready to be merged, we're watching many integrations, take your time and give us a know when ready for merge

**Comment by ctriant** [comment]
> I like this feature, thank you guys! @ctriant let us know when this PR would be ready to be merged, we're watching many integrations, take your time and give us a know when ready for merge

@peppelinux i think we are ready, i integrated the suggestions of @nsklikas

File: src/oidcop/oidc/token.py (hunk 3 above)
Comment (peppelinux) [reply src/oidcop/oidc/token.py]: ok @nsklikas I wait for your revision before merge

**Review by nsklikas — APPROVED** [review]

**Pull request merged**